### PR TITLE
Improve help deploy again

### PR DESCRIPTION
**Pull request by pmatulis**
*This is a new PR based on #9344.*

---

## Description of change

Improving the Examples section for the `deploy` command was my original goal here but I ended up going over all the text as well. I cut out one `--device` example as it was for providing general syntax (the documentation should handle that).

There is a lack of consistency in how the Examples section is formatted in many Juju commands. I would like to offer up the Examples section in this PR as a standard. A few commands already use it. Sample:

```
Examples:

Deploy to a new machine:

    juju deploy apache2

Deploy to machine 23:

    juju deploy mysql --to 23

Deploy to a new LXD container on a new machine:

    juju deploy mysql --to lxd
```  

## QA steps

View the help (`juju help deploy`) and see the revised output.

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

**Comment by pmatulis** [comment]
[Improve help for `deploy` command](https://trello.com/c/wZVBxlSk/721-improve-help-for-deploy-command)

File: cmd/juju/application/deploy.go [hunk 1]
```diff
@@ -387,80 +387,85 @@ type DeployCommand struct {
...
 
-For cs:~user/trusty/mysql
-  ~user/mysql
+The second and third forms are using charm URLs.
 
-For cs:bundle/mediawiki-single
-  mediawiki-single
-  bundle/mediawiki-single
+A versioned charm URL will be expanded as expected. For example, 'mysql-33'
+becomes 'cs:xenial/mysql-33'.
```
Comment (kwmonroe) [inline cmd/juju/application/deploy.go]: Minor knit, but mysql-33 doesn't actually exist.

https://jujucharms.com/mysql/33

Let's use `mysql-56` instead, which exists and expands to `cs:xenial/mysql-56`:

https://jujucharms.com/mysql/56

File: cmd/juju/application/deploy.go [hunk 2]
```diff
@@ -387,80 +387,85 @@ type DeployCommand struct {
...
+  juju deploy charm --series xenial --force
 
-  juju deploy /path/to/bundle/openstack/bundle.yaml
+When charms that include LXD profiles are deployed the profiles are validated
+using a hardcoded allow/deny list. This check can be bypassed with the use of
+the '--force' option. However, doing so is not recommended as it can lead to
+unexpected behaviour.
+
+If an 'application name' is not provided, the application name used is the charm
+or bundle name. A user-supplied 'application name' must consist only of
```
Comment (kwmonroe) [inline cmd/juju/application/deploy.go]: `or bundle name.`

That isn't quite right.  When deploying a bundle, the `application name` is ignored afaict.

File: cmd/juju/application/deploy.go [hunk 3]
```diff
@@ -387,80 +387,85 @@ type DeployCommand struct {
...
+
+Use the '--device' option to specify GPU device requirements (with Kubernetes).
+The below format is used for this option's value, where the 'label' is named in
+the charm metadata file:
 
-Devices can be specified by specifying the '--device' option to deploy charms to a
-k8s cluster which require the use of a GPU (or many).
-Devices provided should be in format:
-	<label>=[<count>,]<device-class>|<vendor/type>[,<attributes>]
+  <label>=[<count>,]<device-class>|<vendor/type>[,<attributes>]
```
Comment (kwmonroe) [inline cmd/juju/application/deploy.go]: I'm not seeing a `--device` option in `juju-2.4.4-xenial-amd64`. Should this (and prior para) be removed?

File: cmd/juju/application/deploy.go [hunk 4]
```diff
@@ -469,105 +474,126 @@ use
...
 
-	juju deploy mycharm --device bitcoinminer=nvidia.com/gpu
-	(deploy mycharm requires any Nvidia GPU. No count is specified, it is assumed to be 1)
+Deploy a Kubernetes charm that requires two Nvidia GPUs that have an
+attribute of 'gpu=nvidia-tesla-p100':
 
-	juju deploy mycharm --device bitcoinminer=1,nvidia.com/gpu,gpu=nvidia-tesla-p100;attr2=attr2
-	(deploy mycharm requires 1*nvidia.com/gpu with attributes: gpu=nvidia-tesla-p10 && attr2=attr2)
+    juju deploy mycharm --device \
+       twingpu=2,nvidia.com/gpu,gpu=nvidia-tesla-p100
```
Comment (kwmonroe) [inline cmd/juju/application/deploy.go]: Here's that pesky `--device` thing again.  I must be missing something, but this is what I get on 2.4.4:
```
$ juju deploy mycharm --device miner=1,nvidia.com/gpu
ERROR flag provided but not defined: --device
```

File: cmd/juju/application/deploy.go (hunk 1 above)
Comment (pmatulis) [reply cmd/juju/application/deploy.go]: :heavy_check_mark:

File: cmd/juju/application/deploy.go (hunk 2 above)
Comment (pmatulis) [reply cmd/juju/application/deploy.go]: :heavy_check_mark:

File: cmd/juju/application/deploy.go (hunk 3 above)
Comment (pmatulis) [reply cmd/juju/application/deploy.go]: This PR is for 2.5. For snaps, it is in the --edge channel.

File: cmd/juju/application/deploy.go (hunk 4 above)
Comment (pmatulis) [reply cmd/juju/application/deploy.go]: 2.5 only.

File: cmd/juju/application/deploy.go [hunk 5]
```diff
@@ -387,80 +387,86 @@ type DeployCommand struct {
...
-  mysql
-  trusty/mysql
+  juju deploy postgresql
+  juju deploy xenial/postgresql
+  juju deploy cs:xenial/postgresql
+  juju deploy postgresql --series xenial
 
-For cs:~user/trusty/mysql
-  ~user/mysql
+The second and third forms are using charm URLs.
```
Comment (hmlanigan) [inline cmd/juju/application/deploy.go]: What are the first and fourth forms?  Technically those commands will use the charm store.  Can we have a charm URL not from the charm store?  I don't know the answer to that.  And maybe isn't necessary for this update?

File: cmd/juju/application/deploy.go [hunk 6]
```diff
@@ -387,80 +387,86 @@ type DeployCommand struct {
...
+  juju deploy charm --series xenial --force
 
-  juju deploy /path/to/bundle/openstack/bundle.yaml
+When charms that include LXD profiles are deployed the profiles are validated
+for security purposes by allowing only certain configurations and devices. Use
+the '--force' option to bypass this check. Doing so is not recommended as it
+can lead to unexpected behaviour.
+
+An 'application name' provides an alternate name for the application. It works
+only for charms; it is silently ignored for bundles. Such a name must consist
```
Comment (hmlanigan) [inline cmd/juju/application/deploy.go]: A bundle can have an 'application name' for a charm and it is used.  The openstack bundles rename the percona-cluster charm to mysql.

Comment (pmatulis) [reply cmd/juju/application/deploy.go]: Right. I meant when used with the `deploy` command. You mean using an application name within the bundle itself right?

File: cmd/juju/application/deploy.go (hunk 5 above)
Comment (kwmonroe) [reply cmd/juju/application/deploy.go]: Yeah, looks like all those forms are charm store URLs.  Trying to deploy something ambiguous (local charm is in the pwd) gives you a warning to be more specific:
```
ubuntu@b413deb34dfe:~/charms/xenial$ juju deploy pig
ERROR The charm or bundle "pig" is ambiguous.
To deploy a local charm or bundle, run `juju deploy ./pig`.
To deploy a charm or bundle from the store, run `juju deploy cs:pig`.
```

I think we could drop this sentence about 2nd/3rd forms, or note that all forms will communicate with the charm store.  The warning is pretty clear, and deploying local charms is covered below.

File: cmd/juju/application/deploy.go (hunk 6 above)
Comment (kwmonroe) [reply cmd/juju/application/deploy.go]: Right, app names can be specified in the bundle, but I think this comment was describing the `application name` parameter for the `deploy` command:
```
$ juju deploy --help | head
Usage: juju deploy [options] <charm or bundle> [<application name>]
```

Type whatever you want for that last param when deploying a bundle; it'll be ignored (at least in my quick check).

File: cmd/juju/application/deploy.go (hunk 5 above)
Comment (pmatulis) [reply cmd/juju/application/deploy.go]: I'll drop the sentence. I just felt the need to explain "charm URL" somehow.

**Pull request merged**